File: diploma-server/src/main/java/org/arhor/diploma/web/filter/JwtAuthTokenFilter.java
package org.arhor.diploma.web.filter;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.arhor.diploma.web.security.JwtProvider;
import org.springframework.lang.NonNull;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

@Slf4j
@Component
@RequiredArgsConstructor
public class JwtAuthTokenFilter extends OncePerRequestFilter {

  private final JwtProvider tokenProvider;
  private final UserDetailsService userService;

  @Override
  protected void doFilterInternal(
      @NonNull HttpServletRequest req,
      @NonNull HttpServletResponse res,
      @NonNull FilterChain filterChain) throws ServletException, IOException {
    try {
      String jwt = getJwt(req);
      if (jwt != null && tokenProvider.tokenIsValid(jwt)) {
        String username = tokenProvider.getUserNameFromJwtToken(jwt);

        UserDetails userDetails = userService.loadUserByUsername(username);

        // if (SecurityContextHolder.getContext().getAuthentication() == null) { ... }

        final var auth =
            new UsernamePasswordAuthenticationToken(
                userDetails,
                null,
                userDetails.getAuthorities());

        auth.setDetails(new WebAuthenticationDetailsSource().buildDetails(req));

        SecurityContextHolder.getContext().setAuthentication(auth);
      }
    } catch (Exception e) {
      log.error("Can NOT set user authentication -> Message: {}", e.getMessage());
    }

    filterChain.doFilter(req, res);
  }

  private String getJwt(HttpServletRequest request) {
    final String authHeader = request.getHeader("Authorization");

    if (authHeader != null && authHeader.startsWith("Bearer ")) {
      return authHeader.replace("Bearer ", "");
    }

    return null;
  }
}

File: diploma-server/src/main/kotlin/org/arhor/diploma/web/filter/JwtAuthTokenFilter.kt
package org.arhor.diploma.web.filter

import org.arhor.diploma.service.AccountService
import org.arhor.diploma.web.security.TokenProvider
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.web.authentication.WebAuthenticationDetailsSource
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class JwtAuthTokenFilter(
    private val tokenProvider: TokenProvider<*>,
    private val accountService: AccountService
) : OncePerRequestFilter() {

  companion object {
    @JvmStatic
    private val log: Logger = LoggerFactory.getLogger(JwtAuthTokenFilter::class.java)
  }

  override fun doFilterInternal(
      req: HttpServletRequest,
      res: HttpServletResponse,
      next: FilterChain
  ) {
    try {
      req.getHeader("Authentication")
          ?.let(tokenProvider::parse)
          ?.takeIf(tokenProvider::validate)
          ?.let(tokenProvider::parseUsername)
          ?.let(accountService::loadUserByUsername)
          ?.let { userDetails ->
            val auth =
                UsernamePasswordAuthenticationToken(
                    userDetails,
                    null,
                    userDetails.authorities)
            auth.details = WebAuthenticationDetailsSource().buildDetails(req)
            SecurityContextHolder.getContext().authentication = auth
          }
    } catch (e: Exception) {
      log.error("Can NOT set user authentication -> Message: {}", e.message)
    }
    next.doFilter(req, res)
  }
}
